pretty/marshal.go

```diff
@@ -0,0 +1,232 @@
+	return EncodeConfig{
+		Line:   []byte{'\n'},
+		Indent: []byte{'\t'},
+		Sep:    []byte{},
+	}.Marshal(n, w)
+}
+
+type EncodeConfig struct {
+	Line   []byte // typically: '\n'
+	Indent []byte // typically: '\t'
+	Sep    []byte // typically: ',' or nothing
+
+	// the following limits can be used to configure elision:
+	// (exceeding them causes "..." in the output)
+	// (not yet implemented)
+
+	DepthLimit int // every child beyond this is replaced with "...".
+	CountLimit int // this counter is per depth level, not in total (use TotalLimit for that).
+	TotalLimit int // this counter is the total number of elements emitted.  map start, list start, and each scalar counts as one.  map keys do not count.
+}
+
+func (cfg EncodeConfig) Marshal(n ipld.Node, w io.Writer) error {
+	st := encodeState{cfg: cfg, w: w, count: make([]int, 0, 10)}
+	return st.marshal(n)
+}
+
+type encodeState struct {
+	cfg     EncodeConfig
+	w       io.Writer
+	err     error
+	scratch [64]byte
+	count   []int
+	total   int
+	// depth int // = len(count)
+}
+
+func (st *encodeState) marshal(n ipld.Node) error {
+	st.total++
+	if tn, ok := n.(schema.TypedNode); ok {
+		st.write("(")
+		st.write(string(tn.Type().Name()))
+		st.write(") ")
+	}
+	switch n.ReprKind() {
+	case ipld.ReprKind_Null:
+		if n.IsAbsent() {
+			st.write("absent")
+		} else {
+			st.write("null")
+		}
+		return st.checkErr()
+	case ipld.ReprKind_Map:
+		st.count = append(st.count, 0)
+		st.write("map node {")
+		st.writeBytes(st.cfg.Line)
+		for itr := n.MapIterator(); !itr.Done(); {
+			k, v, err := itr.Next()
+			if err != nil {
+				return err
+			}
+			st.count[len(st.count)-1]++
+			st.writeIndent()
+			if err := st.emitFlat(k); err != nil {
+				return err
+			}
+			st.write(": ")
+			if err := st.marshal(v); err != nil {
+				return err
+			}
+			st.writeBytes(st.cfg.Sep)
+			st.writeBytes(st.cfg.Line)
+			if err := st.checkErr(); err != nil {
+				return err
+			}
+		}
+		st.count = st.count[0 : len(st.count)-1]
+		st.writeIndent()
+		st.write("}")
+		return st.checkErr()
+	case ipld.ReprKind_List:
+		st.count = append(st.count, 0)
+		st.write("list node [")
+		st.writeBytes(st.cfg.Line)
+		for itr := n.ListIterator(); !itr.Done(); {
+			i, v, err := itr.Next()
+			if err != nil {
+				return err
+			}
+			st.count[len(st.count)-1]++
+			st.writeIndent()
+			b := strconv.AppendInt(st.scratch[:0], int64(i), 10)
+			st.writeBytes(b)
+			st.write(": ")
+			if err := st.marshal(v); err != nil {
+				return err
+			}
+			st.writeBytes(st.cfg.Sep)
+			st.writeBytes(st.cfg.Line)
+			if err := st.checkErr(); err != nil {
+				return err
+			}
+		}
+		st.count = st.count[0 : len(st.count)-1]
+		st.writeIndent()
+		st.write("]")
+		return st.checkErr()
+	case ipld.ReprKind_Bool:
+		v, _ := n.AsBool()
+		switch v {
+		case true:
+			st.write("bool node: true")
+			return st.checkErr()
+		case false:
+			st.write("bool node: false")
+			return st.checkErr()
+		default:
```

Comment:
this case is quite literally unreachable though, so you can just remove it

or even better, use an if/else

Reply:
I like to stick to uses switch cases when doing marshal/unmarshal stuff.

But it encounters a problem on bools: the golang compiler doesn't currently seem to recognize that booleans are exhausted after these two cases, and insists on this panic.  If the panic isn't present, the whole function is considered to have a branch that doesn't have a return statement!

I suppose I might as well change to `if`, though, yes.